bittide-rust-ffi/src/lib.rs

```diff
@@ -0,0 +1,530 @@
+    , steady_state_target: 3.185
+    , rf_state: SOME_RF_STATE_WAIT
+    };
+
+#[allow(dead_code)]
+const CASTED_CONTROL_STATE: *const (i32, u32, f32, u32, f32, u32) =
+  &SOME_CONTROL_STATE as *const _ as *const (i32, u32, f32, u32, f32, u32);
+
+const_assert!(
+  54321i32 ==
+    { let (x,_,_,_,_,_) = unsafe {*CASTED_CONTROL_STATE}; x }
+);
+
+const_assert!(
+  1u32 ==
+    { let (_,x,_,_,_,_) = unsafe {*CASTED_CONTROL_STATE}; x }
+);
+
+const_assert!(
+  3.185f32 ==
+    { let (_,_,x,_,_,_) = unsafe {*CASTED_CONTROL_STATE}; x }
+);
+
+const_assert!(
+  2u32 ==
+    { let (_,_,_,x,_,_) = unsafe {*CASTED_CONTROL_STATE}; x }
+);
+
+const_assert!(
+  3.321f32 ==
+    { let (_,_,_,_,x,_) = unsafe {*CASTED_CONTROL_STATE}; x }
+);
+
+const_assert!(
+  12345u32 ==
+    { let (_,_,_,_,_,x) = unsafe {*CASTED_CONTROL_STATE}; x }
+);
+
+/// Produces the same output as the corresponding Haskell `Show`
+/// instance.
+impl fmt::Display for ControlSt {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!
+      ( f
+      , "ControlSt {{\
+         _z_k = {}, \
+         _b_k = {}, \
+         _steadyStateTarget = {}, \
+         rfState = {}\
+         }}"
+      , self.z_k
+      , self.b_k
+      , self.steady_state_target
+      , self.rf_state
+      )
+  }
+}
+
+/// Allows to cast a "`void`" pointer into a mutable `ControlSt`
+/// reference (unsafe).
+impl ReprRust<*const ()> for &mut ControlSt {
+  fn from(ptr: *const ()) -> Self {
+    return unsafe { mem::transmute(ptr) };
+  }
+}
+
+/// Allows to casts a "`void`" pointer into a non-mutable `ControlSt`
+/// reference (unsafe).
+impl ReprRust<*const ()> for &ControlSt {
+  fn from(ptr: *const ()) -> Self {
+    return unsafe { mem::transmute(ptr) };
+  }
+}
+
+/// Rust sibling of
+/// `Bittide.ClockControl.StabilityChecker.StabilityIndication`.
+#[derive(PartialEq, PartialOrd)]
+pub struct StabilityIndication(usize);
+
+impl StabilityIndication {
+  /// Indicates stability of the signal over time.
+  pub fn stable(&self) -> bool {
+    return self.0 & 1 == 1;
+  }
+  /// Indicates whether the signal is stable and close to
+  /// `targetDataCount`.
+  pub fn settled(&self) -> bool {
+    return self.0 & 2 == 2;
+  }
+}
+
+/// Produces the same output as the corresponding Haskell `Show`
+/// instance.
+impl fmt::Display for StabilityIndication {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    write!
+      ( f
+      , "StabilityIndication {{\
+         stable = {}, \
+         settled = {}\
+         }}"
+      , if self.stable()  { "True" } else { "False" }
+      , if self.settled() { "True" } else { "False" }
+      )
+  }
+}
+
+/// Type wrapper for `Vec<StabilityIndication>`.
+pub struct VSI(Vec<StabilityIndication>);
+
+/// Allows to casts a "`void`" pointer into a non-mutable `ControlSt`
+/// reference (unsafe).
+impl ReprRust<*const ()> for &VSI {
+  fn from(ptr: *const ()) -> Self {
+    return unsafe { mem::transmute(ptr) };
+  }
+}
```

Comment:
Since all access to this vector is immutable, it doesn't need to be a `Vec<>` at all, it could be a `&'a [StabilityIndication]` which can be constructed in a stable way with [`core::slice::from_raw_parts()`](https://doc.rust-lang.org/stable/core/slice/fn.from_raw_parts.html), unlike a `Vec` which has an internally unstable layout that just happens to be (ptr, len, cap) for now. The parameter could then be a `(*const (), usize)` instead and then we don't need a `transmute`